File: take_u_forward/graph/35_Cheapest Flights Within K Stops.cpp
class Solution {
public:
    int CheapestFLight(int n, vector<vector<int>>& flights,
                       int src, int dst, int K) {

        // Adjacency list: node -> {neighbor, cost}
        vector<vector<pair<int,int>>> adj(n);
        for (auto &f : flights) {
            adj[f[0]].push_back({f[1], f[2]});
        }

        // Queue format: {stops, {node, cost}}
        queue<pair<int, pair<int,int>>> q;
        q.push({0, {src, 0}});

        vector<int> dist(n, 1e9);
        dist[src] = 0;

        while (!q.empty()) {
            auto cur = q.front();
            q.pop();

            int stops = cur.first;
            int node  = cur.second.first;
            int cost  = cur.second.second;

            if (stops > K) continue;

            for (auto &it : adj[node]) {
                int nextNode = it.first;
                int price    = it.second;

                // Only relax cost if cheaper AND within stop limit
                if (cost + price < dist[nextNode]) {
                    dist[nextNode] = cost + price;
                    q.push({stops + 1, {nextNode, cost + price}});
                }
            }
        }

        return dist[dst] == 1e9 ? -1 : dist[dst];
    }
};

File: take_u_forward/graph/35_Cheapest Flights Within K Stops.java
import java.util.*;

class Solution {

    public int CheapestFLight(int n, int[][] flights, int src, int dst, int k) {

        // Build adjacency list
        List<List<int[]>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());

        for (int[] f : flights) {
            adj.get(f[0]).add(new int[]{f[1], f[2]});
        }

        // Queue: {stops, node, cost}
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0, src, 0});

        int[] dist = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[src] = 0;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int stops = cur[0];
            int node  = cur[1];
            int cost  = cur[2];

            if (stops > k) continue;

            for (int[] edge : adj.get(node)) {
                int nextNode = edge[0];
                int price    = edge[1];

                if (cost + price < dist[nextNode]) {
                    dist[nextNode] = cost + price;
                    q.offer(new int[]{stops + 1, nextNode, cost + price});
                }
            }
        }

        return dist[dst] == Integer.MAX_VALUE ? -1 : dist[dst];
    }
}